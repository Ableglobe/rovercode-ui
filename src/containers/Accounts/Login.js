import React, { Component, Fragment } from 'react';
import { Redirect } from 'react-router';
import {
  Button,
  Form,
  Grid,
  Icon,
  Message,
  Segment,
} from 'semantic-ui-react';
import axios from 'axios';
import URL from 'url-parse';
import queryString from 'query-string';
import PropTypes from 'prop-types';
import { withCookies, Cookies } from 'react-cookie';

class Login extends Component {
  constructor(props) {
    super(props);

    this.state = {
      socialError: false,
      basicError: false,
      basicSuccess: false,
      username: '',
      password: '',
    };
  }

  basicLogin = () => {
    const { cookies } = this.props;
    const { username, password } = this.state;

    return axios.post('/api/api-token-auth/', {
      username,
      password,
    })
      .then((response) => {
        cookies.set('auth_jwt', response.data.token, { path: '/' });
        this.setState({
          basicSuccess: true,
        });
      })
      .catch(() => {
        this.setState({
          basicError: true,
        });
      });
  }

  handlePasswordChange = (event) => {
    this.setState({
      password: event.target.value,
    });
  }

  handleUsernameChange = (event) => {
    this.setState({
      username: event.target.value,
    });
  }

  redirectToSocial = (element) => {
    const service = element.target.id;
    return axios.post(`/jwt/auth/social/${service}/auth-server/`)
      .then((response) => {
        const url = URL(response.data.url);
        const parsed = queryString.parse(url.query);
        const redirectUrl = URL(parsed.redirect_uri);
        redirectUrl.set('pathname', `/login/${service}/callback/`);
        parsed.redirect_uri = redirectUrl.toString();
        url.query = queryString.stringify(parsed);
        window.location.assign(url.toString());
      })
      .catch(() => {
        this.setState({
          socialError: true,
        });
      });
  }

  render() {
    const { basicError, basicSuccess, socialError } = this.state;

    return (
      <Fragment>
        {
          basicSuccess ? (
            <Redirect to="/" />
          ) : (null)
        }
        <Grid centered columns={16}>
          <Grid.Row>
            <p>
              You can sign up / sign in with one of your existing third-party accounts.
            </p>
          </Grid.Row>
          {
            socialError ? (
              <Grid.Row>
                <Message negative>
                  <Message.Header>
                    There was an error initiating social login.
                  </Message.Header>
                  <p>
                    Contact
                    {' '}
                    <a href="mailto:support@rovercode.com">
                      support@rovercode.com
                    </a>
                    {' '}
                    for help with this issue.
                  </p>
                </Message>
              </Grid.Row>
            ) : (null)
          }
          <Grid.Row>
            <Grid.Column width={8}>
              <Button id="google" floated="right" color="red" onClick={this.redirectToSocial}>
                <Icon name="google" />
                Sign in with Google
              </Button>
            </Grid.Column>
            <Grid.Column width={8}>
              <Button id="github" floated="left" color="black" onClick={this.redirectToSocial}>
                <Icon name="github" />
                Sign in with GitHub
              </Button>
            </Grid.Column>
          </Grid.Row>
          <Grid.Row>
            <p>
              Don&apos;t have any of those accounts?
              {' '}
              <a href="/accounts/signup">
                Create a rovercode account.
              </a>
            </p>
          </Grid.Row>
          <Grid.Row>
            <p>
              Already have a rovercode account? Sign in here:
            </p>
          </Grid.Row>
          {
            basicError ? (
              <Grid.Row>
                <Message negative>
                  <Message.Header>
                    Invalid username or password.
                  </Message.Header>
                </Message>
              </Grid.Row>
            ) : (null)
          }
          <Grid.Row>
            <Grid.Column width={4}>
              <Segment raised secondary>
                <Form floated="left" onSubmit={this.basicLogin}>
                  <Form.Input required icon="user" iconPosition="left" placeholder="Username" onChange={this.handleUsernameChange} />
                  <Form.Input required icon="lock" iconPosition="left" type="password" placeholder="Password" onChange={this.handlePasswordChange} />
                  <Form.Button primary type="submit">
Sign In
                  </Form.Button>
                </Form>
              </Segment>
            </Grid.Column>
          </Grid.Row>
          <Grid.Row>
            <p>
              <a href="/accounts/reset">
                Forgot Password?
              </a>
            </p>
          </Grid.Row>
        </Grid>
      </Fragment>
    );
  }
}

Login.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
};

export default withCookies(Login);
