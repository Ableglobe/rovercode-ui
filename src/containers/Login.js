import React, { Component, Fragment } from 'react';
import {
  Button,
  Grid,
  Icon,
  Image,
  Message,
} from 'semantic-ui-react';
import axios from 'axios';
import URL from 'url-parse';
import queryString from 'query-string';

import logoImage from '../assets/images/rovercode_logo.png';

export default class Login extends Component {
  constructor(props) {
    super(props);

    this.state = {
      error: false,
    };
  }

  redirectToSocial = (element) => {
    const service = element.target.id;
    return axios.post(`/jwt/auth/social/${service}/auth-server/`)
      .then((response) => {
        const url = URL(response.data.url);
        const parsed = queryString.parse(url.query);
        const redirectUrl = URL(parsed.redirect_uri);
        redirectUrl.set('pathname', `/login/${service}/callback/`);
        parsed.redirect_uri = redirectUrl.toString();
        url.query = queryString.stringify(parsed);
        window.location.assign(url.toString());
      })
      .catch(() => {
        this.setState({
          error: true,
        });
      });
  }

  render() {
    const { error } = this.state;

    return (
      <Fragment>
        <Grid centered columns="equal">
          <Grid.Row>
            <Image src={logoImage} />
          </Grid.Row>
          <Grid.Row>
            <p>
              You can sign up / sign in with one of your existing third-party accounts.
            </p>
          </Grid.Row>
          {
            error ? (
              <Grid.Row>
                <Message negative>
                  <Message.Header>
                    There was an error initiating social login.
                  </Message.Header>
                  <p>
                    Contact
                    {' '}
                    <a href="mailto:support@rovercode.com">
                      support@rovercode.com
                    </a>
                    {' '}
                    for help with this issue.
                  </p>
                </Message>
              </Grid.Row>
            ) : (null)
          }
          <Grid.Row>
            <Grid.Column>
              <Button id="google" floated="right" color="red" onClick={this.redirectToSocial}>
                <Icon name="google" />
                Sign in with Google
              </Button>
            </Grid.Column>
            <Grid.Column>
              <Button id="github" floated="left" color="black" onClick={this.redirectToSocial}>
                <Icon name="github" />
                Sign in with GitHub
              </Button>
            </Grid.Column>
          </Grid.Row>
        </Grid>
      </Fragment>
    );
  }
}
