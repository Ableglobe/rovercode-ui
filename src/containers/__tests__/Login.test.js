import React from 'react';
import { Message } from 'semantic-ui-react';
import { shallow } from 'enzyme';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import Login from '../Login';

const mock = new MockAdapter(axios);

test('Login renders on the page with no errors', () => {
  const location = {
    pathname: '/login',
  };
  const wrapper = shallow(<Login location={location} />);

  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find(Message).exists()).toBe(false);
});

test('Login redirects to social api on button click', async () => {
  const url = 'https://accounts.google.com/o/oauth2/auth?redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Fauth%2Fsocial%2Fgoogle%2Fcallback';
  mock.reset();
  mock.onPost('/jwt/auth/social/google/auth-server/').reply(200, { url });
  window.location.assign = jest.fn();

  const location = {
    pathname: '/login',
  };
  const element = {
    target: {
      id: 'google',
    },
  };

  const wrapper = shallow(<Login location={location} />);

  await wrapper.instance().redirectToSocial(element);

  expect(window.location.assign).toBeCalledWith(
    'https://accounts.google.com/o/oauth2/auth?redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Flogin%2Fgoogle%2Fcallback%2F',
  );
});

test('Login shows error message on api error', async () => {
  mock.reset();
  mock.onPost('/jwt/auth/social/google/auth-server/').timeout();
  window.location.assign = jest.fn();

  const location = {
    pathname: '/login',
  };
  const element = {
    target: {
      id: 'google',
    },
  };

  const wrapper = shallow(<Login location={location} />);

  await wrapper.instance().redirectToSocial(element);
  wrapper.update();

  expect(wrapper.find(Message).exists()).toBe(true);
  expect(window.location.assign).not.toBeCalled();
});
