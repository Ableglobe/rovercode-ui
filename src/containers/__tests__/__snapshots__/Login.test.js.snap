// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders on the page with no errors 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    location={
      Object {
        "pathname": "/login",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Grid
        centered={true}
        columns="equal"
      >
        <GridRow>
          <Image
            as="img"
            src={Object {}}
            ui={true}
          />
        </GridRow>
        <GridRow>
          <p>
            You can sign up / sign in with one of your existing third-party accounts.
          </p>
        </GridRow>
        <GridRow>
          <GridColumn>
            <Button
              as="button"
              color="red"
              floated="right"
              id="google"
              onClick={[Function]}
              role="button"
            >
              <Icon
                as="i"
                name="google"
              />
              Sign in with Google
            </Button>
          </GridColumn>
          <GridColumn>
            <Button
              as="button"
              color="black"
              floated="left"
              id="github"
              onClick={[Function]}
              role="button"
            >
              <Icon
                as="i"
                name="github"
              />
              Sign in with GitHub
            </Button>
          </GridColumn>
        </GridRow>
      </Grid>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "centered": true,
        "children": Array [
          <GridRow>
            <Image
              as="img"
              src={Object {}}
              ui={true}
            />
          </GridRow>,
          <GridRow>
            <p>
              You can sign up / sign in with one of your existing third-party accounts.
            </p>
          </GridRow>,
          null,
          <GridRow>
            <GridColumn>
              <Button
                as="button"
                color="red"
                floated="right"
                id="google"
                onClick={[Function]}
                role="button"
              >
                <Icon
                  as="i"
                  name="google"
                />
                Sign in with Google
              </Button>
            </GridColumn>
            <GridColumn>
              <Button
                as="button"
                color="black"
                floated="left"
                id="github"
                onClick={[Function]}
                role="button"
              >
                <Icon
                  as="i"
                  name="github"
                />
                Sign in with GitHub
              </Button>
            </GridColumn>
          </GridRow>,
        ],
        "columns": "equal",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Image
              as="img"
              src={Object {}}
              ui={true}
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "as": "img",
              "src": Object {},
              "ui": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <p>
              You can sign up / sign in with one of your existing third-party accounts.
            </p>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "You can sign up / sign in with one of your existing third-party accounts.",
            },
            "ref": null,
            "rendered": "You can sign up / sign in with one of your existing third-party accounts.",
            "type": "p",
          },
          "type": [Function],
        },
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <GridColumn>
                <Button
                  as="button"
                  color="red"
                  floated="right"
                  id="google"
                  onClick={[Function]}
                  role="button"
                >
                  <Icon
                    as="i"
                    name="google"
                  />
                  Sign in with Google
                </Button>
              </GridColumn>,
              <GridColumn>
                <Button
                  as="button"
                  color="black"
                  floated="left"
                  id="github"
                  onClick={[Function]}
                  role="button"
                >
                  <Icon
                    as="i"
                    name="github"
                  />
                  Sign in with GitHub
                </Button>
              </GridColumn>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Button
                  as="button"
                  color="red"
                  floated="right"
                  id="google"
                  onClick={[Function]}
                  role="button"
                >
                  <Icon
                    as="i"
                    name="google"
                  />
                  Sign in with Google
                </Button>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "as": "button",
                  "children": Array [
                    <Icon
                      as="i"
                      name="google"
                    />,
                    "Sign in with Google",
                  ],
                  "color": "red",
                  "floated": "right",
                  "id": "google",
                  "onClick": [Function],
                  "role": "button",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "as": "i",
                      "name": "google",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "Sign in with Google",
                ],
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Button
                  as="button"
                  color="black"
                  floated="left"
                  id="github"
                  onClick={[Function]}
                  role="button"
                >
                  <Icon
                    as="i"
                    name="github"
                  />
                  Sign in with GitHub
                </Button>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "as": "button",
                  "children": Array [
                    <Icon
                      as="i"
                      name="github"
                    />,
                    "Sign in with GitHub",
                  ],
                  "color": "black",
                  "floated": "left",
                  "id": "github",
                  "onClick": [Function],
                  "role": "button",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "as": "i",
                      "name": "github",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "Sign in with GitHub",
                ],
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Grid
          centered={true}
          columns="equal"
        >
          <GridRow>
            <Image
              as="img"
              src={Object {}}
              ui={true}
            />
          </GridRow>
          <GridRow>
            <p>
              You can sign up / sign in with one of your existing third-party accounts.
            </p>
          </GridRow>
          <GridRow>
            <GridColumn>
              <Button
                as="button"
                color="red"
                floated="right"
                id="google"
                onClick={[Function]}
                role="button"
              >
                <Icon
                  as="i"
                  name="google"
                />
                Sign in with Google
              </Button>
            </GridColumn>
            <GridColumn>
              <Button
                as="button"
                color="black"
                floated="left"
                id="github"
                onClick={[Function]}
                role="button"
              >
                <Icon
                  as="i"
                  name="github"
                />
                Sign in with GitHub
              </Button>
            </GridColumn>
          </GridRow>
        </Grid>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "centered": true,
          "children": Array [
            <GridRow>
              <Image
                as="img"
                src={Object {}}
                ui={true}
              />
            </GridRow>,
            <GridRow>
              <p>
                You can sign up / sign in with one of your existing third-party accounts.
              </p>
            </GridRow>,
            null,
            <GridRow>
              <GridColumn>
                <Button
                  as="button"
                  color="red"
                  floated="right"
                  id="google"
                  onClick={[Function]}
                  role="button"
                >
                  <Icon
                    as="i"
                    name="google"
                  />
                  Sign in with Google
                </Button>
              </GridColumn>
              <GridColumn>
                <Button
                  as="button"
                  color="black"
                  floated="left"
                  id="github"
                  onClick={[Function]}
                  role="button"
                >
                  <Icon
                    as="i"
                    name="github"
                  />
                  Sign in with GitHub
                </Button>
              </GridColumn>
            </GridRow>,
          ],
          "columns": "equal",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Image
                as="img"
                src={Object {}}
                ui={true}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "as": "img",
                "src": Object {},
                "ui": true,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <p>
                You can sign up / sign in with one of your existing third-party accounts.
              </p>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "You can sign up / sign in with one of your existing third-party accounts.",
              },
              "ref": null,
              "rendered": "You can sign up / sign in with one of your existing third-party accounts.",
              "type": "p",
            },
            "type": [Function],
          },
          null,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <GridColumn>
                  <Button
                    as="button"
                    color="red"
                    floated="right"
                    id="google"
                    onClick={[Function]}
                    role="button"
                  >
                    <Icon
                      as="i"
                      name="google"
                    />
                    Sign in with Google
                  </Button>
                </GridColumn>,
                <GridColumn>
                  <Button
                    as="button"
                    color="black"
                    floated="left"
                    id="github"
                    onClick={[Function]}
                    role="button"
                  >
                    <Icon
                      as="i"
                      name="github"
                    />
                    Sign in with GitHub
                  </Button>
                </GridColumn>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <Button
                    as="button"
                    color="red"
                    floated="right"
                    id="google"
                    onClick={[Function]}
                    role="button"
                  >
                    <Icon
                      as="i"
                      name="google"
                    />
                    Sign in with Google
                  </Button>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "as": "button",
                    "children": Array [
                      <Icon
                        as="i"
                        name="google"
                      />,
                      "Sign in with Google",
                    ],
                    "color": "red",
                    "floated": "right",
                    "id": "google",
                    "onClick": [Function],
                    "role": "button",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "as": "i",
                        "name": "google",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    "Sign in with Google",
                  ],
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <Button
                    as="button"
                    color="black"
                    floated="left"
                    id="github"
                    onClick={[Function]}
                    role="button"
                  >
                    <Icon
                      as="i"
                      name="github"
                    />
                    Sign in with GitHub
                  </Button>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "as": "button",
                    "children": Array [
                      <Icon
                        as="i"
                        name="github"
                      />,
                      "Sign in with GitHub",
                    ],
                    "color": "black",
                    "floated": "left",
                    "id": "github",
                    "onClick": [Function],
                    "role": "button",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "as": "i",
                        "name": "github",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    "Sign in with GitHub",
                  ],
                  "type": [Function],
                },
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
